charon-ml/src/PrintUtils.ml

```diff
@@ -26,27 +26,16 @@ type ('fun_body, 'global_body) fmt_env = {
           indices, i.e., the region var of index 0 doesn't have to be at index 0,
           etc. We lookup the variables by checking their id, not their position.
       *)
-  types : type_var list;
-  const_generics : const_generic_var list;
-  trait_clauses : trait_clause list;
-  preds : predicates;
+  generics : generic_params;
+      (* We ignore `generics.regions` as they are tracked in `regions` already *)
```

Comment:
I'm tempted to also remove `regions` and store a stack of `generic_params` instead. This might be a bit too general but I've heard rumors of the new trait solver being able to support `for<T: Clone> X: Trait<Item = T>` kinds of bounds, which would require an actual stack of `generic_params`.

Reply:
It would make sense I think.

Reply:
I beg to differ, any simplification of this whole thing is a huge win at this point

Reply:
I'm about to add methods to match up and substitute `GenericParams` and `GenericArgs`. If we unify here we'd be able to reuse all these methods and checks.

Reply:
On further thought, I think we will need this to handle GATs too.